crates/subspace-networking/src/shared.rs

```diff
@@ -82,15 +83,29 @@ pub(crate) enum Command {
     Dial {
         address: Multiaddr,
     },
+    ConnectedPeers {
+        result_sender: oneshot::Sender<Vec<PeerId>>,
+    },
 }
 
 pub(crate) type HandlerFn<A> = Arc<dyn Fn(&A) + Send + Sync + 'static>;
 type Handler<A> = Bag<HandlerFn<A>, A>;
 
+/// [`PeerInfo`] update and related data container.
+pub struct NewPeerInfo {
+    /// Peer ID for this [`PeerInfo`] update.
+    pub peer_id: PeerId,
+    /// [`PeerInfo`] update.
+    pub peer_info: PeerInfo,
+    /// Currently connected peers.
+    pub connected_peers: Vec<PeerId>,
```

Comment:
This field doesn't belong here. New peer info doesn't mean peer set changed, if you need to know that, there should be a separate event for it.

Reply:
My point was that we don't need to collect anything at all. We simply need to add an event that is fired when peer is connected/disconnected, so we can react to it rather than collecting list of all peers every time we have new peer info. Then update logic that you have simplified will be even simpler.

Reply:
If we create a new event we'll introduce a new handler to support, and can possibly get much more events to process. The current event aggregates information and is invoked only when we need information. I'm not sure that we'll benefit from the proposed change.